virttest/utils_disk.py

```diff
@@ -361,65 +363,554 @@ def create_partition_table(session, did, labeltype, ostype):
+            if binary:
+                factor = 1024
+                dst_unit = order[order.index(unit.replace('iB', '')) - 1]
+                unit = dst_unit + 'iB'
+            else:
+                factor = 1000
+                dst_unit = order[order.index(unit.replace('B', '')) - 1]
+                unit = dst_unit
+
+            size_value = utils_numeric.normalize_data_size(size_value + unit,
+                                                           dst_unit, factor)
+            start_value = utils_numeric.normalize_data_size(start_value + unit,
+                                                            dst_unit, factor)
+
+        while num:
+            end_value = str(float(start_value) + float(size_value))
+            if num == 1:
+                unit = get_misc_info_parted(session, did, binary).unit
+                free = get_partitions_parted(session, did,
+                                             unit, False).free_spaces
+                free_end_value = free[-1]['end'].replace(unit, '')
+
+                # The location end will be outside of the device
+                # for binary unit, if the requested location
+                # is end of free space. So convert it to non-binary unit.
+                if float(end_value) >= float(free_end_value):
+                    end_unit = get_misc_info_parted(session, did, False).unit
+                    free = get_partitions_parted(session, did,
+                                                 end_unit, False).free_spaces
+                    end_value = free[-1]['end'].replace(end_unit, '')
+                    force_sector = False
+                else:
+                    end_unit = unit
+
+                name, partition, _ = create_parted(
+                    session, did, part_type, start_value + unit,
+                    end_value + end_unit, unit, timeout, debug,
+                    force_sector, correct)
+            else:
+                name, partition, _ = create_parted(
+                    session, did, part_type, start_value + unit,
+                    end_value + unit, unit, timeout, debug,
+                    force_sector, correct)
+                start_value = partition['end'].replace(unit, '')
+
+            names.append(name)
+            partitions.append(partition)
+            num -= 1
+    return names, partitions
+
+
+def list_disk_diskpart(session, did, timeout=60, debug=True):
+    """
+    list selected disk info by diskpart in windows.
 
+    :param session: session object to guest.
+    :param did: string type, disk index. e.g: 1, 2.
+    :param timeout: float type, imeout for cmd execution in seconds.
+    :param debug: boolean type, if is true, will display the info of
+    created partition.
+    :return dictionary type, the disk space(status, size, free, dyn, gpt) of
+    selected disk.
+    """
+    disk_script = '_'.join(
+        ("disk", random.sample(string.ascii_letters + string.digits, 4)))
```

Comment:
hi, str type and list type cannot be join together here, will report "TypeError: sequence item 1: expected str instance, list found".
you can change to disk_script = "disk_" + ''.join(random.sample(string.ascii_letters + string.digits, 4)) as other place used.

Reply:
Hi, Peixiu
Updated it, please help to review, thanks.